cirq-google/cirq_google/workflow/quantum_executable.py

```diff
@@ -0,0 +1,157 @@
+from cirq import NamedTopology, _compat
+from cirq.protocols import dataclass_json_dict
+from cirq.study.resolver import _is_param_resolver_or_similar_type
+
+
+class ExecutableSpec(metaclass=abc.ABCMeta):
+    """Specification metadata about an executable.
+
+    Subclasses should add problem-specific fields.
+    """
+
+    executable_family: str = NotImplemented
+    """A unique name to group executables."""
+
+
+@dataclass(frozen=True)
+class BitstringsMeasurement:
+    """Use in-circuit MeasurementGate to collect many repetitions of strings of bits.
+
+    This is the lowest-level measurement type allowed in `QuantumExecutable` and behaves
+    identically to the `cirq.Sampler.run` function. The executable's circuit must contain
+    explicit measurement gates.
+
+    Args:
+        n_repeitions: The number of repetitions to execute the circuit.
+    """
+
+    n_repetitions: int
+
+    def _json_dict_(self):
+        return dataclass_json_dict(self, namespace='cirq.google')
+
+    def __repr__(self):
+        return _compat.dataclass_repr(self, namespace='cirq_google')
+
+
+TParamPair = Tuple[cirq.TParamKey, cirq.TParamVal]
+
+
+@dataclass(frozen=True)
+class QuantumExecutable:
+    """An executable quantum program.
+
+    This serves a similar purpose to `cirq.Circuit` with some key differences. First, a quantum
+    executable contains all the relevant context for execution including parameters as well as
+    the desired number of repetitions. Second, this object is immutable. Finally, there are
+    optional fields enabling a higher level of abstraction for certain aspects of the executable.
+
+    Attributes:
+        circuit: A circuit describing the quantum operations to execute.
+        measurement: A description of the type of measurement. A valid option is to use
+            MeasurementGate in your circuit and specify
+            `measurement=BitstringsMeasurement(n_repetitions)`. Future changes may permit
+            different types of measurement.
+        params: An immutable version of cirq.ParamResolver represented as a tuple of key value
+            pairs.
+        spec: Specification metadata about this executable that is not used by the quantum runtime,
+            but is persisted in result objects to associate executables with results.
+        problem_topology: Description of the multiqubit gate topology present in the circuit.
+            If not specified, the circuit must be compatible with the device topology.
+        initial_state: How to initialize the quantum system before running `circuit`. If not
+            specified, the device will be initialized into the all-zeros state.
+    """
+
+    circuit: cirq.FrozenCircuit
+    measurement: BitstringsMeasurement
+    params: Optional[Tuple[TParamPair, ...]] = None
+    spec: Optional[ExecutableSpec] = None
+    problem_topology: Optional[NamedTopology] = None
+    initial_state: Optional[cirq.ProductState] = None
+
+    # pylint: disable=missing-raises-doc
```

Comment:
Nit: given how many ways this can raise Errors, it's probably worth documenting the raises cases.

Reply:
not done pending clarity on how we should be validating input, see below

Reply:
Now there's only one `ValueError`, which is only raised if the documented public API is not followed, which the style guide advocates against documenting